exporter/datadogexporter/traces_exporter.go

```diff
@@ -16,25 +16,92 @@ package datadogexporter
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.uber.org/zap"
 )
 
 type traceExporter struct {
-	logger *zap.Logger
+	logger         *zap.Logger
+	cfg            *Config
+	edgeConnection apm.TraceEdgeConnection
+	obfuscator     *obfuscate.Obfuscator
+	tags           []string
 }
 
-func newTraceExporter(_ *Config, logger *zap.Logger) (*traceExporter, error) {
-	return &traceExporter{logger}, nil
+func newTraceExporter(logger *zap.Logger, cfg *Config) (*traceExporter, error) {
+	// removes potentially sensitive info and PII, approach taken from serverless approach
+	// https://github.com/DataDog/datadog-serverless-functions/blob/11f170eac105d66be30f18eda09eca791bc0d31b/aws/logs_monitoring/trace_forwarder/cmd/trace/main.go#L43
+	obfuscator := obfuscate.NewObfuscator(&obfuscate.Config{
+		ES: obfuscate.JSONSettings{
+			Enabled: true,
+		},
+		Mongo: obfuscate.JSONSettings{
+			Enabled: true,
+		},
+		RemoveQueryString: true,
+		RemovePathDigits:  true,
+		RemoveStackTraces: true,
+		Redis:             true,
+		Memcached:         true,
+	})
+
+	// Calculate tags at startup
+	tags := cfg.TagsConfig.GetTags(false)
+	// TODO:
+	// use passed in config values for site and api key instead of hardcoded
+	exporter := &traceExporter{
+		logger:         logger,
+		cfg:            cfg,
+		edgeConnection: apm.CreateTraceEdgeConnection(cfg.Traces.TCPAddr.Endpoint, cfg.API.Key, false),
+		obfuscator:     obfuscator,
+		tags:           tags,
+	}
+
+	return exporter, nil
 }
 
 func (exp *traceExporter) pushTraceData(
-	_ context.Context,
+	ctx context.Context,
 	td pdata.Traces,
 ) (int, error) {
 
-	// TODO: add implementation
+	// convert traces to datadog traces and group trace payloads by env
+	// we largely apply the same logic as the serverless implementation, simplified a bit
+	// https://github.com/DataDog/datadog-serverless-functions/blob/f5c3aedfec5ba223b11b76a4239fcbf35ec7d045/aws/logs_monitoring/trace_forwarder/cmd/trace/main.go#L61-L83
+	ddTraces, err := convertToDatadogTd(td, exp.cfg, exp.tags)
+
+	if err != nil {
+		exp.logger.Info(fmt.Sprintf("Failed to convert traces with error %v\n", err))
+		return 0, err
+	}
+
+	// group the traces by env to reduce the number of flushes
+	aggregatedTraces := aggregateTracePayloadsByEnv(ddTraces)
+
+	// security/obfuscation for db, query strings, stack traces, pii, etc
+	// TODO: is there any config we want here? OTEL has their own pipeline for regex obfuscation
+	apm.ObfuscatePayload(exp.obfuscator, aggregatedTraces)
+
+	for _, ddTracePayload := range aggregatedTraces {
+
+		err := exp.edgeConnection.SendTraces(context.Background(), ddTracePayload, 3)
+
+		if err != nil {
+			exp.logger.Info(fmt.Sprintf("Failed to send traces with error %v\n", err))
+		}
+
+		// this is for generating metrics like hits, errors, and latency, it uses a seperate endpoint than Traces
+		stats := apm.ComputeAPMStats(ddTracePayload)
+		err_stats := exp.edgeConnection.SendStats(context.Background(), stats, 3)
```

Comment:
i believe duping 1/2 would be a possibility here, the endpoint is pretty naive iirc. There's a note in the serverless repo on this. 

Thinking about it though, it probably makes sense to prioritize `SendTraces` imo, since metrics are based on the traces, so if there's something unaccepted with the traces we probably don't want to trust the metrics that are generated anyway

Reply:
Okay, guess we will have to deal with it :( I still think we could benefit from having a function like

 ```go
func pushWithRetry(maxRetries int, waitTime time.Duration, fn func() error) error {
  // ...
}
```

that abstracts the retrying logic in case we want to change it afterwards, but feel free to mark this as done if you think that's not worth the effort.

Reply:
yea i think that's reasonable, let me see if i can make this a bit cleaner

Reply:
I tried to add at least the structure of how we'd use a retry  func here. wiring up all the wait Duration retry stuff seemed a bit complicated so i passed on that for the moment, but i think we now at least have the structure where we can just update the error handling function if we want to handle retries differently.

I also dropped the retry amount to 0 (passing in 1), since it's unclear to me if we dedupe in most cases, so would rather not experiment, if we experience lots of drops we can iterate.